### ISSUE TITLE
job queing in SJS

### ISSUE THREAD

**Issue by yiu31802** [spark-jobserver/spark-jobserver#789]
Hello,

because simultaneous job triggers cause SJS server down in my environment, I desire a job queuing mechanism in SJS. However, I got to learn about this StakOverflow with @noorul 

http://stackoverflow.com/questions/34145979/spark-jobserver-max-jobs-per-context

Is this still valid? Is there no plan onwards in this regard? Please share some technical difficulty if any. Not really sure but I may want to invest time on this implementation...

Thanks,

Junji

**Comment by amarouni** [comment]
@yiu31802 Yeah there's no queuing mechanism in SJS and from experience it's not an easy task. Spark + separate JVM per context is a recipe for heavy resource usage. Maybe that's something we can start looking at as a community because it looks like a requested feature.

@noorul @velvia What do you think ? Can this feature be something that we can work on next ?

**Comment by velvia** [comment]
@amarouni yeah this has been requested for a long time.  The challenge is designing it and scope.   A lot of it depends on the architecture of your system, and implementing it means we choose one opinion.  :)  

I think context-per-ivm is not just a matter of resources, it’s also about isolation and a necessity for certain workloads.  It’s also more in line with how Spark was designed.

**Comment by amarouni** [comment]
@velvia A simple pluggable queuing module can resolve some of the issues as mentioned by OP. A queue that sits in front of the job execution endpoints (/job & /context)  can help probably.

Another feature that's requested regularly is the SJS High availability, which can help with the overloading of the SJS, but I'll leave that for another discussion.

In the meantime I think we can come up with a simple job/context queue built in SJS, for limiting the number of concurrent jobs/contexts started. @velvia what do you think ?

**Comment by velvia** [comment]
I’m fine with a simple queue.  I really thought there was another issue to track this, where i had put down much more design thoughts, but can’t find it now  :/

Anyways, each Akka mailbox is already a queue.  I wonder if we couldn’t take advantage of it.   However I guess it’s not the same thing, the most common use case for a queue would be if the number of slots or threads is already full, thus other jobs have to wait.  For this we do need a modification.

**Comment by yiu31802** [comment]
Hello, I just implemented a quick-and-dirty job queuing mechanism in javascript: https://www.npmjs.com/package/sjs-wrapper I actually use this package cause our SJS server needs to load balance the job request from many users.
